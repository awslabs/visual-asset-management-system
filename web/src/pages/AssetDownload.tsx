import { useLocation } from "react-router";
import { Storage } from "aws-amplify";
import { FileTree } from "../components/filemanager/FileManager";
import { FileUploadTable, FileUploadTableItem } from "./AssetUpload/FileUploadTable";
import { useReducer, useState } from "react";

async function downloadAllFilesRecursively(tree: FileTree, directoryHandle: any, dispatch: any) {
    for (let subtree of tree.subTree) {
        if (subtree.subTree.length > 0) {
            const subFolderDirectoryHandle = await directoryHandle.getDirectoryHandle(
                subtree.name,
                { create: true }
            );
            await downloadAllFilesRecursively(subtree, subFolderDirectoryHandle, dispatch);
        } else {
            const fileHandle = await directoryHandle.getFileHandle(subtree.name, { create: true });
            const writable = await fileHandle.createWritable();
            const data = await Storage.get(subtree.keyPrefix, {
                download: true,
                progressCallback: (progress: any) => {
                    console.log("Progress callback ", progress);
                    dispatch({
                        type: "UPDATE_PROGRESS",
                        payload: {
                            status: "In Progress",
                            relativePath: subtree.relativePath,
                            progress: progress.loaded,
                            loaded: progress.loaded,
                            total: progress.total,
                        },
                    });
                },
            });
            await writable.write(data.Body);
            await writable.close();
            dispatch({
                type: "UPDATE_STATUS",
                payload: { relativePath: subtree.relativePath, status: "Completed" },
            });
        }
    }
}

async function downloadFolder(tree: FileTree, dispatch: any) {
    //@ts-ignore
    const directoryHandle = await window.showDirectoryPicker();
    await downloadAllFilesRecursively(tree, directoryHandle, dispatch);
}
const convertFileTreeItemsToFileUploadTableItems = (fileTree: FileTree): FileUploadTableItem[] => {
    const allItems: FileUploadTableItem[] = [];
    for (let subtree of fileTree.subTree) {
        if (subtree.subTree.length > 0) {
            allItems.push(...convertFileTreeItemsToFileUploadTableItems(subtree));
        } else {
            allItems.push({
                name: subtree.name,
                index: 0,
                size: 0,
                relativePath: subtree.relativePath,
                status: "Queued",
                progress: 0,
                startedAt: Date.now(),
                loaded: 0,
                total: 0,
            });
        }
    }
    return allItems;
};

const updateIndices = (fileUploadTableItems: FileUploadTableItem[]): FileUploadTableItem[] => {
    let index = 0;
    for (let item of fileUploadTableItems) {
        item.index = index;
        index++;
    }
    return fileUploadTableItems;
};

function assetDownloadReducer(
    state: FileUploadTableItem[],
    action: { type: string; payload: any }
): FileUploadTableItem[] {
    switch (action.type) {
        case "UPDATE_INDICES":
            return updateIndices(state);
        case "UPDATE_PROGRESS":
            return state.map((item) => {
                if (item.relativePath === action.payload.relativePath) {
                    return {
                        ...item,
                        status: action.payload.status,
                        size: action.payload.total,
                        progress: Math.floor((action.payload.loaded / action.payload.total) * 100),
                        loaded: action.payload.loaded,
                        total: action.payload.total,
                    };
                } else {
                    return item;
                }
            });
        case "UPDATE_STATUS":
            return state.map((item) => {
                if (item.relativePath === action.payload.relativePath) {
                    return {
                        ...item,
                        status: action.payload.status,
                    };
                } else {
                    return item;
                }
            });
        default:
            return state;
    }
}
export default function AssetDownloadsPage() {
    const { state } = useLocation();
    const fileTree = state["fileTree"] as FileTree;
    const [resume, setResume] = useState(true);
    const fileUploadTableItems = convertFileTreeItemsToFileUploadTableItems(fileTree);
    const [fileUploadTableItemsState, dispatch] = useReducer(
        assetDownloadReducer,
        fileUploadTableItems
    );
    return (
        <div>
            <h1> Downloading {fileTree.relativePath} </h1>
            <FileUploadTable
                allItems={fileUploadTableItemsState}
                resume={resume}
                onRetry={() => {
                    downloadFolder(fileTree, dispatch).then(() => {
                        setResume(false);
                    });
                }}
                mode={"Download"}
            />
        </div>
    );
}
