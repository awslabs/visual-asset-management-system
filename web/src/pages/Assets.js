/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { useNavigate } from "react-router-dom";
import { fetchAllAssets, fetchDatabaseAssets } from "../services/APIService";
import { AssetListDefinition } from "../components/list/list-definitions/AssetListDefinition";
import ListPage from "./ListPage";
import { useParams } from "react-router";
import Synonyms from "../synonyms";

export default function Assets() {
    const navigate = useNavigate();

    const urlParams = useParams();
    return (
        <ListPage
            singularName={Synonyms.Asset}
            singularNameTitleCase={Synonyms.Asset}
            pluralName={Synonyms.assets}
            pluralNameTitleCase={Synonyms.Assets}
            onCreateCallback={() => {
                if (urlParams.databaseId) {
                    navigate(`/upload/${urlParams.databaseId}`);
                } else {
                    navigate("/upload");
                }
            }}
            listDefinition={AssetListDefinition}
            fetchAllElements={fetchAllAssets}
            fetchElements={fetchDatabaseAssets}
        />
    );
}
