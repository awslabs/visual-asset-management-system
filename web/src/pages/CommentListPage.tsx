/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from "react";

import {
    Box,
    BreadcrumbGroup,
    Button,
    Grid,
    SpaceBetween,
    TextContent,
} from "@cloudscape-design/components";
import { useParams } from "react-router";
import CommentTableList from "../components/list/CommentTableList";
import PropTypes, { object } from "prop-types";
import ListDefinition from "../components/list/list-definitions/types/ListDefinition";
import RelatedTableList from "../components/list/RelatedTableList";
import Synonyms from "../synonyms";

export default function CommentListPage(props: any) {
    const { databaseId } = useParams();
    const {
        singularNameTitleCase,
        pluralName,
        pluralNameTitleCase,
        listDefinition,
        CreateNewElement,
        fetchElements,
        fetchAllElements,
        onCreateCallback,
        isRelatedTable,
        editEnabled,
        onSelection,
        selectedItems,
    } = props;
    const [reload, setReload] = useState<boolean>(true);
    const [loading, setLoading] = useState<boolean>(true);
    const [allItems, setAllItems] = useState<Array<any>>([]);

    const [openNewElement, setOpenNewElement] = useState(false);

    useEffect(() => {
        const getData = async () => {
            setLoading(true);
            let items;
            if (databaseId) {
                items = await fetchElements({ databaseId: databaseId });
            } else {
                items = await fetchAllElements();
            }

            if (items !== false && Array.isArray(items)) {
                setLoading(false);
                setReload(false);
                setAllItems(
                    //@todo fix workflow delete return
                    items.filter((item) => item.databaseId.indexOf("#deleted") === -1)
                );
            }
        };
        if (reload) {
            getData();
        }
    }, [reload, databaseId, fetchAllElements, fetchElements]);

    const handleOpenNewElement = () => {
        if (onCreateCallback) onCreateCallback();
        else if (CreateNewElement) setOpenNewElement(true);
    };

    return (
        <>
            <Box padding={{ top: databaseId ? "s" : "m", horizontal: "l" }}>
                {databaseId && (
                    <BreadcrumbGroup
                        items={[
                            { text: Synonyms.Databases, href: "/databases/" },
                            {
                                text: databaseId,
                                href: `/databases/${databaseId}/${pluralName}/`,
                            },
                        ]}
                        ariaLabel="Breadcrumbs"
                    />
                )}
                <Grid gridDefinition={[{ colspan: { default: 6 } }]}>
                    <div>
                        <TextContent>
                            <h1>
                                {pluralNameTitleCase}
                                {databaseId && ` for ${databaseId}`}
                            </h1>
                        </TextContent>
                    </div>
                </Grid>
                <Grid gridDefinition={[{ colspan: { default: 12 } }]}>
                    {isRelatedTable && (
                        <RelatedTableList
                            allItems={allItems}
                            loading={loading}
                            listDefinition={listDefinition}
                            databaseId={databaseId}
                            setReload={setReload}
                        />
                    )}
                    <CommentTableList
                        allItems={allItems}
                        loading={loading}
                        listDefinition={listDefinition}
                        databaseId={databaseId}
                        editEnabled={editEnabled}
                        setReload={setReload}
                        onSelection={onSelection}
                        selectedItems={selectedItems}
                        UpdateSelectedElement={CreateNewElement}
                        createNewElement={
                            (CreateNewElement || onCreateCallback) && (
                                <div style={{ float: "right" }}>
                                    <SpaceBetween direction={"horizontal"} size={"m"}>
                                        <Button onClick={handleOpenNewElement} variant="primary">
                                            Create {singularNameTitleCase}
                                        </Button>
                                    </SpaceBetween>
                                </div>
                            )
                        }
                    />
                </Grid>
            </Box>
            {CreateNewElement && (
                <CreateNewElement
                    open={openNewElement}
                    setOpen={setOpenNewElement}
                    setReload={setReload}
                    databaseId={databaseId}
                />
            )}
        </>
    );
}

CommentListPage.propTypes = {
    singularName: PropTypes.string.isRequired,
    singularNameTitleCase: PropTypes.string.isRequired,
    pluralName: PropTypes.string.isRequired,
    pluralNameTitleCase: PropTypes.string.isRequired,
    listDefinition: PropTypes.instanceOf(ListDefinition).isRequired,
    CreateNewElement: PropTypes.func,
    fetchElements: PropTypes.func.isRequired,
    fetchAllElements: PropTypes.func,
    onCreateCallback: PropTypes.func,
    onSelection: PropTypes.func,
    selectedItems: PropTypes.arrayOf(object),
    isRelatedTable: PropTypes.bool,
    editEnabled: PropTypes.bool,
};
