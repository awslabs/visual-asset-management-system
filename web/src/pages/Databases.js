/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from "react";
import { fetchAllDatabases } from "../services/APIService";
import CreateDatabase from "../components/createupdate/CreateDatabase";
import { DatabaseListDefinition } from "../components/list/list-definitions/DatabaseListDefinition";
import ListPage from "./ListPage";

import { anyRoleOf } from "../FedAuth/roles";

export default function Databases() {
    const [createDatabase, setCreateDatabase] = useState(false);
    useEffect(() => {
        anyRoleOf(["create-database", "super-admin"]).then((result) => {
            setCreateDatabase(result);
        });
    }, []);
    return (
        <ListPage
            singularName={"database"}
            singularNameTitleCase={"Database"}
            pluralName={"databases"}
            pluralNameTitleCase={"Databases"}
            listDefinition={DatabaseListDefinition}
            CreateNewElement={(createDatabase && CreateDatabase) || undefined}
            fetchAllElements={fetchAllDatabases}
            fetchElements={fetchAllDatabases}
            editEnabled={true}
        />
    );
}
