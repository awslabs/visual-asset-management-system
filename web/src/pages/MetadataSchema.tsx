/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import {
    Box,
    Button,
    Checkbox,
    FileUpload,
    Form,
    FormField,
    Grid,
    Input,
    Modal,
    Multiselect,
    MultiselectProps,
    Select,
    SelectProps,
    SpaceBetween,
    TextContent,
} from "@cloudscape-design/components";
import { createContext, useContext, useState } from "react";
import { generateUUID } from "../common/utils/utils";

import { API, Storage } from "aws-amplify";

import ListDefinition from "../components/list/list-definitions/types/ListDefinition";
import ColumnDefinition from "../components/list/list-definitions/types/ColumnDefinition";
import ListPageNoDatabase from "./ListPageNoDatabase";
import { NonCancelableEventHandler } from "@cloudscape-design/components/internal/events";
import { useNavigate, useParams } from "react-router";
import DatabaseSelectorWithModal from "../components/selectors/DatabaseSelectorWithModal";

export interface SchemaContextData {
    schemas: MetadataSchemaFields[];
    databaseId: string | null;
}

const SchemaContext = createContext<SchemaContextData>({ schemas: [], databaseId: null });

interface CreateMetadataFieldProps {
    open: boolean;
    setOpen: (open: boolean) => void;
    setReload: (reload: boolean) => void;
    initState: any;
}

type DataTypes =
    | "string"
    | "textarea"
    | "number"
    | "boolean"
    | "date"
    | "location"
    | "controlled-list"
    | "inline-controlled-list";
export interface MetadataSchemaFields {
    id: string;
    field: string;
    dataType: DataTypes;
    required: boolean;
    databaseId: string;
    sequenceNumber: number;
    dependsOn: string[];
    // location fields
    useCenterPosition?: boolean;
    latitudeField?: string;
    longitudeField?: string;
    zoomLevelField?: string;

    // inline controlled list fields
    inlineControlledListValues?: string;
}

interface DatatypeSelectProps {
    value: string;
    onChange: NonCancelableEventHandler<SelectProps.ChangeDetail>;
    disabled: boolean;
}
function DatatypeSelect({ value, onChange, disabled }: DatatypeSelectProps) {
    const options: SelectProps.Option[] = [
        { value: "string", label: "Text" },
        { value: "textarea", label: "Multiline Text Area" },
        { value: "number", label: "Number" },
        { value: "boolean", label: "Boolean" },
        { value: "date", label: "Date" },
        { value: "location", label: "Location" },
        { value: "controlled-list", label: "Controlled List" },
        { value: "inline-controlled-list", label: "Inline Controlled List" },
    ];
    return (
        <Select
            selectedOption={options.find((x) => x.value === value) || null}
            disabled={disabled}
            onChange={onChange}
            options={options}
        />
    );
}

interface DependencyMultiselectProps {
    selectedOptions: string[];
    onChange: NonCancelableEventHandler<MultiselectProps.MultiselectChangeDetail>;
    disabled?: boolean;
    api?: typeof API;
}

function DependencyMultiselect({
    selectedOptions,
    onChange,
    disabled,
    api = API,
}: DependencyMultiselectProps) {
    const { schemas } = useContext(SchemaContext);

    return (
        <Multiselect
            options={schemas.map((x) => ({ value: x.field, label: x.field }))}
            selectedOptions={
                selectedOptions && selectedOptions.map((x) => ({ value: x, label: x }))
            }
            onChange={onChange}
            disabled={!(schemas?.length > 0) || disabled}
        />
    );
}

function validateField(field: string | null | undefined) {
    let response = null;
    if (!field) {
        response = "Field is required";
    }
    if (field && field.length < 4) {
        response = "Field must be at least 4 characters";
    }
    return response;
}

function validateDataType(dataType: string) {
    let response = null;
    if (!dataType) {
        response = "Data Type is required";
    }
    return response;
}

interface LocationSpecificFieldsProps {
    formState: MetadataSchemaFields;
    setFormState: (formState: MetadataSchemaFields) => void;
}

function InlineControlledListSpecificFields({
    formState,
    setFormState,
}: LocationSpecificFieldsProps) {
    if (formState.dataType !== "inline-controlled-list") {
        return null;
    }

    return (
        <>
            <FormField
                label="Inline Controlled List Values"
                constraintText="Define a list of values that can be selected from a dropdown."
                errorText={validateField(formState.inlineControlledListValues)}
            >
                <Input
                    value={formState.inlineControlledListValues!}
                    onChange={({ detail }) => {
                        setFormState({
                            ...formState,
                            inlineControlledListValues: detail.value,
                        });
                    }}
                />
            </FormField>
        </>
    );
}

function LocationSpecificFields({ formState, setFormState }: LocationSpecificFieldsProps) {
    if (formState.dataType !== "location") {
        return null;
    }

    if (!formState.useCenterPosition) {
        return (
            <FormField
                label="Use Center Position"
                constraintText="Use the center point of the map as the initial position."
            >
                <Checkbox
                    checked={formState.useCenterPosition || false}
                    onChange={({ detail }) => {
                        setFormState({
                            ...formState,
                            useCenterPosition: detail.checked,
                        });
                    }}
                />
            </FormField>
        );
    }

    return (
        <>
            <FormField
                label="Use Center Position"
                constraintText="Define a center point from the metadata."
            >
                <Checkbox
                    checked={formState.useCenterPosition}
                    onChange={({ detail }) => {
                        setFormState({
                            ...formState,
                            useCenterPosition: detail.checked,
                        });
                    }}
                />
            </FormField>
            <FormField
                label="Latitude Field"
                constraintText="The field from which to initialize the center point of the map."
                errorText={validateField(formState.latitudeField)}
            >
                <Input
                    value={formState.latitudeField || ""}
                    onChange={({ detail }) => {
                        setFormState({
                            ...formState,
                            latitudeField: detail.value,
                        });
                    }}
                />
            </FormField>
            <FormField
                label="Longitude Field"
                constraintText="The field from which to initialize the center point of the map."
                errorText={validateField(formState.longitudeField)}
            >
                <Input
                    value={formState.longitudeField || ""}
                    onChange={({ detail }) => {
                        setFormState({
                            ...formState,
                            longitudeField: detail.value,
                        });
                    }}
                />
            </FormField>
            <FormField
                label="Zoom Field"
                constraintText="The field from which to initialize the zoom level of the map."
                errorText={validateField(formState.zoomLevelField)}
            >
                <Input
                    value={formState.zoomLevelField || ""}
                    onChange={({ detail }) => {
                        setFormState({
                            ...formState,
                            zoomLevelField: detail.value,
                        });
                    }}
                />
            </FormField>
        </>
    );
}

function CreateMetadataField({ open, setOpen, setReload, initState }: CreateMetadataFieldProps) {
    const createOrUpdate = (initState && "Update") || "Create";

    const [inProgress, setInProgress] = useState(false);
    const { databaseId } = useContext(SchemaContext);
    const [formState, setFormState] = useState<MetadataSchemaFields>({
        id: generateUUID(),
        required: false,
        dataType: "string",
        ...initState,
    });

    return (
        <Modal
            visible={open}
            onDismiss={() => {
                setOpen(false);
            }}
            size="medium"
            header={`${createOrUpdate} Metadata Field`}
            footer={
                <Box float="right">
                    <SpaceBetween direction="horizontal" size="xs">
                        <Button variant="link" onClick={() => setOpen(false)}>
                            Cancel
                        </Button>

                        <Button
                            variant="primary"
                            onClick={() => {
                                setInProgress(true);
                                console.log("sending state", formState);
                                formState.required = formState.required ? true : false;
                                API.post("api", `metadataschema/${databaseId}`, {
                                    // API.post("api", `metadataschema/all`, {
                                    body: formState,
                                })
                                    .then((res) => {
                                        setOpen(false);
                                        setReload(true);
                                        setFormState({
                                            id: generateUUID(),
                                            ...initState,
                                        });
                                    })
                                    .catch((err) => {
                                        console.log("create auth criteria error", err);
                                    })
                                    .finally(() => {
                                        setInProgress(false);
                                    });
                            }}
                            disabled={
                                inProgress ||
                                !(
                                    validateField(formState.field) === null &&
                                    validateDataType(formState.dataType) === null
                                )
                            }
                            data-testid={`${createOrUpdate}-metadataschema-button`}
                        >
                            {createOrUpdate} Field
                        </Button>
                    </SpaceBetween>
                </Box>
            }
        >
            <Form>
                <FormField
                    label="Field"
                    constraintText="The name of the field. Cannot be changed after the field is created."
                    errorText={validateField(formState.field)}
                >
                    <Input
                        value={formState.field}
                        disabled={createOrUpdate === "Update"}
                        onChange={({ detail }) => {
                            setFormState({
                                ...formState,
                                field: detail.value,
                            });
                        }}
                    />
                </FormField>
                <FormField
                    label="Data Type"
                    constraintText="The data type of the field. Cannot be changed after the field is created."
                    errorText={validateDataType(formState.dataType)}
                >
                    <DatatypeSelect
                        disabled={createOrUpdate === "Update"}
                        value={formState.dataType}
                        onChange={({ detail }) => {
                            if (detail.selectedOption.value) {
                                setFormState({
                                    ...formState,
                                    dataType: detail.selectedOption.value as DataTypes,
                                });
                            }
                        }}
                    />
                </FormField>
                <LocationSpecificFields formState={formState} setFormState={setFormState} />
                <InlineControlledListSpecificFields
                    formState={formState}
                    setFormState={setFormState}
                />
                <FormField label="Required" constraintText="Whether the field is required.">
                    <Select
                        selectedOption={
                            formState.required
                                ? { value: "true", label: "Required" }
                                : { value: "false", label: "Not Required" }
                        }
                        onChange={({ detail }) => {
                            if (detail.selectedOption.value) {
                                setFormState({
                                    ...formState,
                                    required: detail.selectedOption.value === "true",
                                });
                            }
                        }}
                        options={[
                            { value: "true", label: "Required" },
                            { value: "false", label: "Not Required" },
                        ]}
                    />
                </FormField>
                <FormField
                    label="Dependencies"
                    constraintText="Fields that must be selected before this field is selected."
                >
                    <DependencyMultiselect
                        selectedOptions={formState.dependsOn}
                        onChange={({ detail }) => {
                            const selected = detail.selectedOptions.map((x: any) => x.value);
                            setFormState({
                                ...formState,
                                dependsOn: selected,
                            });
                        }}
                    />
                </FormField>
                <FormField
                    label="Sequence Number"
                    constraintText="The order in which fields are displayed. Lower values are displayed first."
                >
                    <Input
                        value={formState?.sequenceNumber?.toString()}
                        type="number"
                        onChange={({ detail }) => {
                            setFormState({
                                ...formState,
                                sequenceNumber: Number.parseInt(detail.value),
                            });
                        }}
                    />
                </FormField>
                {/*
                <FormField
                    label="Database"
                    constraintText="The database that this field belongs to. Cannot be changed after the field is created."
                >
                    <DatabaseSelector
                        selectedOption={{
                            label: formState.databaseId,
                            value: formState.databaseId,
                        }}
                        onChange={({ detail }: any) => {
                            const selectedOption = detail.selectedOption;
                            console.log("selectedOption", selectedOption);
                            if (selectedOption && selectedOption.value) {
                                setFormState({
                                    ...formState,
                                    databaseId: selectedOption.value,
                                });
                            }
                        }}
                    />
                </FormField>
                */}
            </Form>
        </Modal>
    );
}

// In order to validate set of fields, wrap the list page in a
// context provider and set the list from the fetch all each time
// it is called so that it can be validated and leveraged in the
// create field form.

export const MetadataSchemaListDefinition = new ListDefinition({
    pluralName: "fields",
    pluralNameTitleCase: "Fields",
    visibleColumns: ["field", "dataType", "required", "sequenceNumber"],
    filterColumns: [{ name: "field", placeholder: "Field" }],
    elementId: "id",
    deleteFunction: async (item: any): Promise<[boolean, string]> => {
        try {
            console.log("delete item", item);
            const response: any = await API.del(
                "api",
                `metadataschema/${item.databaseId}/${item.field}`,
                {}
            );
            return [true, response.message];
        } catch (error: any) {
            console.log(error);
            return [false, error?.message];
        }
    },
    columnDefinitions: [
        new ColumnDefinition({
            id: "field",
            header: "Field",
            cellWrapper: (props: any) => {
                return <>{props.children}</>;
            },
            sortingField: "field",
        }),
        // sequenceNumber
        new ColumnDefinition({
            id: "sequenceNumber",
            header: "Sequence Number",
            cellWrapper: (props: any) => <>{props.children ? props.children : ""}</>,
            sortingField: "sequenceNumber",
        }),
        new ColumnDefinition({
            id: "dataType",
            header: "Data Type",
            cellWrapper: (props: any) => <>{props.children}</>,
            sortingField: "dataType",
        }),
        new ColumnDefinition({
            id: "required",
            header: "Required",
            cellWrapper: (props: any) => <>{props.children}</>,
            sortingField: "required",
        }),
    ],
});

class ProgressCallbackArgs {
    loaded!: number;
    total!: number;
}

async function uploadAssetToS3(
    file: File,
    key: string,
    metadata: { [k: string]: string },
    progressCallback: (progress: ProgressCallbackArgs) => void
) {
    console.log("upload", key, file);
    return Storage.put(key, file, { metadata, progressCallback });
}

function ControlledListFileUpload() {
    const { databaseId } = useParams();
    const [file, setFile] = useState<File[]>([]);

    const schemaKey = `metadataschema/${databaseId}/controlledlist.csv`;

    // form that takes a single file upload and places the file in s3 using Storage.put.
    // the file is then uploaded to the database's s3 bucket.

    return (
        <Form>
            <FormField label="File Upload">
                <FileUpload
                    accept=".csv"
                    multiple={false}
                    value={file}
                    i18nStrings={{
                        uploadButtonText: (e) => (e ? "Choose files" : "Choose file"),
                        dropzoneText: (e) => (e ? "Drop files to upload" : "Drop file to upload"),
                        removeFileAriaLabel: (e) => `Remove file ${e + 1}`,
                        limitShowFewer: "Show fewer files",
                        limitShowMore: "Show more files",
                        errorIconAriaLabel: "Error",
                    }}
                    onChange={({ detail }) => {
                        setFile(detail.value);
                        uploadAssetToS3(
                            detail.value[0],
                            schemaKey,
                            {},
                            (progress: ProgressCallbackArgs) => {
                                console.log("progress", progress);
                            }
                        ).then((result: any) => {
                            console.log("result", result);
                        });
                        console.log("file upload", detail);
                    }}
                />
            </FormField>
        </Form>
    );
}

export default function MetadataSchema() {
    const { databaseId } = useParams();
    const navigate = useNavigate();
    const [databaseSelectModelOpen, setDatabaseSelectModelOpen] = useState(
        databaseId === undefined
    );

    const [schemas, setSchemas] = useState<MetadataSchemaFields[]>([]);
    async function fetchAll(api = API) {
        const resp = await api.get("api", `metadataschema/${databaseId}`, {});
        if (resp.schemas) {
            // sort resp.schemas wrt sequenceNumber
            // undefined values go last
            resp.schemas.sort((a: any, b: any) => {
                if (a.sequenceNumber === undefined) {
                    return 1;
                }
                if (b.sequenceNumber === undefined) {
                    return -1;
                }
                return a.sequenceNumber - b.sequenceNumber;
            });
            setSchemas(resp.schemas);
            return resp.schemas;
        } else {
            return false;
        }
    }

    if (!databaseId) {
        return (
            <DatabaseSelectorWithModal
                open={databaseSelectModelOpen}
                setOpen={setDatabaseSelectModelOpen}
                onSelectorChange={(event: any) => {
                    const id = event?.detail?.selectedOption?.value;
                    navigate(`/metadataschema/${id}/create`);
                }}
            />
        );
    }

    return (
        <SchemaContext.Provider value={{ schemas, databaseId }}>
            <Box padding={{ top: "m", horizontal: "l" }}>
                <Grid gridDefinition={[{ colspan: 6 }]}>
                    <div>
                        <TextContent>
                            <h1>Controlled List Dataset Upload</h1>
                        </TextContent>
                        <ControlledListFileUpload />
                    </div>
                </Grid>
            </Box>
            <ListPageNoDatabase
                singularName={"Metadata Schema Field"}
                singularNameTitleCase={"Metadata Schema Field"}
                pluralName={"Metadata Schema Fields"}
                pluralNameTitleCase={"Metadata Schema Fields"}
                listDefinition={MetadataSchemaListDefinition}
                editEnabled={true}
                CreateNewElement={CreateMetadataField}
                fetchAllElements={fetchAll}
                fetchElements={fetchAll}
            />
        </SchemaContext.Provider>
    );
}
