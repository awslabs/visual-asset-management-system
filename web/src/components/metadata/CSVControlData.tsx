/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { SchemaContextData } from "../../pages/MetadataSchema";
import Papa, { ParseRemoteConfig, ParseResult } from "papaparse";
import React from "react";

export const hashCode = (x: string) => {
    let hash = 0;
    for (let i = 0; i < x.length; i++) {
        hash = (hash << 5) - hash + x.charCodeAt(i);
        hash |= 0; // Convert to 32bit integer
    }
    return hash;
};

export function createPapaParseConfig(
    schema: SchemaContextData,
    setControlledLists: React.Dispatch<any>,
    setRawControlData: React.Dispatch<any>
) {
    const config: ParseRemoteConfig = {
        download: true,
        header: true,
        complete: (results: ParseResult<any>) => {
            const lists = schema.schemas
                .filter((field) => field.dataType === "controlled-list")
                .map((field) => {
                    const deps = field.dependsOn ? field.dependsOn : [];
                    const def: { [k: number]: any } = {};
                    return {
                        field: field.field,
                        columns: [field.field, ...deps],
                        data: def,
                    };
                });

            for (const result of results.data) {
                for (const list of lists) {
                    const item = list.columns.reduce((acc: any, column: string) => {
                        acc[column] = result[column];
                        return acc;
                    }, {});
                    const hash = hashCode(JSON.stringify(item));
                    list.data[hash] = item;
                }
            }
            const listsSorted = lists.map((x) => {
                return {
                    ...x,
                    data: Object.values(x.data).sort((a, b) => {
                        if (JSON.stringify(a) < JSON.stringify(b)) return -1;
                        if (JSON.stringify(a) > JSON.stringify(b)) return 1;
                        return 0;
                    }),
                };
            });
            const byField: { [k: string]: any } = {};
            setControlledLists(
                listsSorted.reduce((acc, x) => {
                    acc[x.field] = x;
                    return acc;
                }, byField)
            );
            setRawControlData(results.data);
        },
    };
    return config;
}

export type HandleControlData = (
    url: string,
    schema: SchemaContextData,
    setControlledLists: React.Dispatch<any>,
    setRawControlData: React.Dispatch<any>
) => void;

export const handleCSVControlData: HandleControlData = function handleCSVControlData(
    url: string,
    schema: SchemaContextData,
    setControlledLists: React.Dispatch<any>,
    setRawControlData: React.Dispatch<any>
) {
    // note: config has a callback with a side effect that calls setControlledLists
    const config: ParseRemoteConfig = createPapaParseConfig(
        schema,
        setControlledLists,
        setRawControlData
    );
    Papa.parse(url, config);
};
