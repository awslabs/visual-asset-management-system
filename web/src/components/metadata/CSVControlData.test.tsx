/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import Papa, { ParseRemoteConfig, ParseResult, ParseWorkerConfig } from "papaparse";
import { createPapaParseConfig, hashCode } from "./CSVControlData";
import { SchemaContextData } from "../../pages/MetadataSchema";

const fs = require("fs");
const path = require("path");

describe("CSVControlData", () => {
    test("hashCode", () => {
        expect(hashCode("a string to hash")).toBe(260647971);
        expect(hashCode("another string to hash")).toBe(-1152987839);
    });

    test("handleCSVControlData", () => {
        // schema has the fields country_name, state_name
        const schema: SchemaContextData = {
            databaseId: "adatabaseid",
            schemas: [
                {
                    id: "122",
                    databaseId: "adatabaseid",
                    field: "country_name",
                    dataType: "controlled-list",
                    dependsOn: [],
                    required: false,
                    sequenceNumber: 0,
                },
                {
                    databaseId: "adatabaseid",
                    field: "state_name",
                    dataType: "controlled-list",
                    dependsOn: ["country_name"],
                    id: "123",
                    required: false,
                    sequenceNumber: 0,
                },
            ],
        };
        let data = null;
        const setControlledLists = jest.fn((x) => {
            data = x;
        });
        const setRawControlData = jest.fn();
        const config: ParseRemoteConfig = createPapaParseConfig(
            schema,
            setControlledLists,
            setRawControlData
        );
        const config2: ParseWorkerConfig = {
            worker: true,
            complete: (results: ParseResult<any>) => {
                config.complete(results, "");
            },
        };

        // read contents of cities.metadata.test.csv into a string
        const citiesMetadataTestCSV = fs.readFileSync(
            path.join(__dirname, "cities.metadata.test.csv"),
            "utf8"
        );

        Papa.parse(citiesMetadataTestCSV, config2);

        expect(setControlledLists).toHaveBeenCalledTimes(1);
        expect(setRawControlData).toHaveBeenCalledTimes(1);
        expect(data).toMatchSnapshot(data);
    });
});
