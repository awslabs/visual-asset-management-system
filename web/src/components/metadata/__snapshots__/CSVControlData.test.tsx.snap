// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSVControlData handleCSVControlData 1`] = `
Object {
  "country_name": Object {
    "columns": Array [
      "country_name",
    ],
    "data": Array [
      Object {
        "country_name": undefined,
      },
    ],
    "field": "country_name",
  },
  "state_name": Object {
    "columns": Array [
      "state_name",
      "country_name",
    ],
    "data": Array [
      Object {
        "country_name": undefined,
        "state_name": undefined,
      },
    ],
    "field": "state_name",
  },
}
`;
